Notebook Cells: Show a visual indicator that the cell has been edited

martinRenou opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/master/CONTRIBUTING.md
-->

## References

One way to fix #6790

Please tell me what you think

## User-facing changes

![edit](https://user-images.githubusercontent.com/21197331/119665405-a756ea00-be34-11eb-913d-1e32237f39e1.gif)

telamonian commented:
That seems almost too subtle an effect for a "invalidated"/"stale" marker. I think it would make sense to at least experiment with also fading the "thumb" on the left side of the active cell from blue -> orange, in addition to still fading the cell execution counter from blue -> orange like you currently do. That would be a simple and consistent way to increase the "volume" of the effect.

In essence, my thinking is that the people who want this feature probably also want it to act like a kind of hazard sign. In general hazard signs should be loud and easily noticed

goanpeca commented:
Pretty cool @martinRenou!

I am bit concerned with accessibility for this feature, since color changes might not be enough? When a file is edited you usually see a `*` or similar. Would something like that provide a cue that is more than a color change?

```
   [1]: 
  *[1]:
```

I put in the left since that way we could preserve some space in there for it? 

Question: Since this is taking into account changes since last execution is this also taking into account if the cell is back to the original content? In that case doing some backspaces would make it non dirty again?

---

Pinging @isabela-pf @telamonian @tonyfast

martinRenou commented:
> That seems almost too subtle an effect for a "invalidated"/"stale" marker. I think it would make sense to at least experiment with also fading the "thumb" on the left side of the active cell from blue -> orange, in addition to still fading the cell execution counter from blue -> orange like you currently do. That would be a simple and consistent way to increase the "volume" of the effect.

Do you mean the vertical bar? If so I totally agree.

martinRenou commented:
> I am bit concerned with accessibility for this feature, since color changes might not be enough? When a file is edited you usually see a * or similar. Would something like that provide a cue that is more than a color change?

fcollonval commented:
This is a great idea @martinRenou 

I wonder if we should also add a `isDirty` attribute on the code cell. So extension could use the status to display more information in the status bar or prompt for a re-run.

On the design side, should a visual element be displayed on the output to highlight it is dirty?

telamonian commented:
> Do you mean the vertical bar?

@martinRenou Yep, that's exactly what I meant. It's similar in shape to the "thumb" of a scrollbar, so I figure "thumb" is a reasonable thing to call this bit of our UI flare

goanpeca commented:
>Although this might bring confusion with the "save" state of the Notebook. What do you think?

It might, but then it is at the cell level not at the tab level, so I think it is clear enough 🤷🏼

martinRenou commented:
@goanpeca do you think we should use both an orange color and a star?

goanpeca commented:
>@goanpeca do you think we should use both an orange color and a star?

I have seen the use of both in some editors, so my initial feeling would be yes.

krassowski commented:
> That seems almost too subtle an effect for a "invalidated"/"stale" marker. I think it would make sense to at least experiment with also fading the "thumb" on the left side of the active cell from blue -> orange, in addition to still fading the cell execution counter from blue -> orange like you currently do. That would be a simple and consistent way to increase the "volume" of the effect.
> 
> In essence, my thinking is that the people who want this feature probably also want it to act like a kind of hazard sign. In general hazard signs should be loud and easily noticed

This is a prime idea but please be careful with modifying `jp-Collapser` colour - it is already being used by nbsafety https://github.com/nbsafety-project/nbsafety for related but subtly different purpose of highlighting outdated cells based on the actual cell dependencies inferred by specialized kernel:

![Screenshot from 2021-05-26 14-50-18](https://user-images.githubusercontent.com/5832902/119671630-d3bd3700-be31-11eb-9f92-e8741433e0a2.png)

The solution proposed by @goanpeca of adding a start/circle indicator + modifying colour seems more compatible with nbsafety; I am not saying that we should not modify the `jp-Collapser` a.k.a. tumb a.k.a vertical bar, but let's coordinate with the wider community.

Hi @smacke, @ruiduoray, @shreyashankar you might be interested in this discussion and in a very good position to contribute your insight from the nbsafety work.

martinRenou commented:
@goanpeca @telamonian  Taking some of your comments into account

![edit2](https://user-images.githubusercontent.com/21197331/119676410-1edd4700-be3e-11eb-89ff-76863695199f.gif)

@fcollonval I added an `isDirty` getter as you suggested

@telamonian concerning your speed concern, I implemented the callback in a way that it would evaluate `isDirty` only once, and only if the cell has been executed once. We might still want to debounce it.

goanpeca commented:
@martinRenou that is looking awesome. 🚀

goanpeca reviewed (approved):
🚀

telamonian commented:
Looks great. The only thing I don't love in the latest gif is the reuse of the asterisk `*` to mean both "dirty" and then later on "running". I get that you're changing it's position when it's meaning is supposed to change, but I don't think we should engage in asterisk semaphore here. We should instead either use a different char (eg dagger `†`, tilde `~`, cdot `•`, empty-circle `⃝`, etc) or just drop the char decorator entirely for the dirty state

smacke commented:
@krassowski Thanks for the shout-out! In my opinion, both approaches (collapser or circle indicator) seem solid and I don't see a strong argument for one over the other. Re. conflicting with nbsafety -- in some sense the visual indicators from nbsafety are only relevant for unedited cells, so from my viewpoint it's on us to make sure that a "cell edited" indicator doesn't get overriden by nbsafety. Also we were planning to move to highlighting / underlining actual code snippets instead of the kitchen-sink approach of an indicator for the entire cell at some point.

davidbrochart commented:
This is a nice improvement! Just a couple of comments.
It makes me wonder if a dirty cell indicator should even be part of the Notebook format. Today it is possible to run a code cell that produces an output, change the cell input, and save it. It kind of goes against reproducibility, because if I open the notebook a rerun the cell, I won't get the same output.
I saw that you don't mark the cell as dirty if only heading/trailing spaces are changed, because this wouldn't change the cell execution. AFAIK this holds true for any programming language, but it kind of goes against the expected behavior of a text editor, which marks the file as dirty as soon as you insert/remove a character, so I'm not 100% sure we should keep that. Also does it remove the dirty flag if I undo my changes?
For the dirty indicator, `vim` uses the `+` character, maybe that would be preferable to `*` which is also used when the cell is executing? Or maybe it is time to use something like a spinner for the cell execution, but I don't know how hard it would be:
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/4711805/119832733-bdc97800-befe-11eb-87bd-64921a02a0a8.gif)

martinRenou commented:
I changed it to a dot:
![dot](https://user-images.githubusercontent.com/21197331/119834056-e9992d80-beff-11eb-8e28-3d3d0c51722d.gif)

What do you think? I personally like it better. I am not a big fan of the spinner idea, having infinitely moving parts in the Notebook UI would be too much of a distraction.

davidbrochart commented:
I also prefer the dot, thanks!

martinRenou commented:
> but it kind of goes against the expected behavior of a text editor, which marks the file as dirty as soon as you insert/remove a character

I understand your point, although this dirty flag is really Notebook specific and does not have the same meaning as in classic text editors. It's dirty because the code does not reflect the output, not because it's not saved.

martinRenou commented:
> It makes me wonder if a dirty cell flag should even be part of the Notebook format

That would make a lot of sense IMO

krassowski commented on packages/cells/src/widget.ts:
```diff
@@ -955,13 +968,30 @@ export class CodeCell extends Cell<ICodeCellModel> {
+    ) {
+      this._isDirty = true;
+      this.addClass(DIRTY_CLASS);
+    }
```
Can we make it a round-trip so that the status is cleared if the user accidentally enters a character but then reverts the change? For example:

```suggestion
    } else if (
      this._isDirty &&
      this._executedCode == model.value.text.trim()
    ) {
      this._isDirty = false;
      this.removeClass(DIRTY_CLASS);
    }
```

telamonian replied:
That does seem like a nice UX to have. For example, I've always appreciated how smart git (and transitively most of the IDEs that integrate with it) is about recognizing when a user has done a "manual" revert.

Again though, my concern is that this potentially very large check is occurring on every user input in a cell. It's not likely to be a problem in 95% of use cases, but it's pretty easy to imagine things going wrong in the case of very large cell input. One not-particularly-nice usage pattern that I see a lot of (especially in notebooks made by recently minted quants) is the copy and paste dump of the entire contents of a CSV, which easily top `10^6` chars. I'm not saying that the "large literal dump" is good practice, this kind of usage already causes some severe performance issues. But I am concerned about performance in general, and also about making a bad situation worse, especially for users who are new to jupyter and/or python itself.

krassowski replied:
I think that it is extremely cheap in comparison to rendering `10^6` of characters on screen. If you do as much as enter a new line, the CodeMirror will have to reflow literally millions of nodes. I would not worry about this check too much.

telamonian replied:
You probably have a point about the relative cost of rendering. I do still feel like I need a little convincing that this is GRAS w.r.t. performance, as I *really* hate having to deal with sluggish keyboard input in code tools. I dunno. Maybe the daily trauma of having to code in a remote VM for work is getting to me.

Also, relevant to nothing, that's a slight exaggeration of the number of nodes/reflows; if they at least manage to get the csv into one string then it's only two nodes (a `<span>` nested in a `<pre>`) per line. Typically codemirror will only create a new node on the same line if/when it gets to a point where the syntax highlighting color need to change. But yes still terrible

martinRenou replied:
@telamonian what would you think of implementing what @krassowski is suggesting here, but debouncing the call?

krassowski replied:
This one seems not resolved. 

@telamonian yes, a milion nodes is an overestimate, very much depends on whether it is syntax highlighted or wrapped into a string. Though is this really a use case that we want to support? From my perspective this an abuse of notebooks, and I would not want to have an annoying feature (orange colouring of cell that has not changed) because the system was designed with such behaviour in mind. This is different than having a long notebook or large output - those are definitely valid use cases (which we sometimes fail to support well).

krassowski commented:
Merge and split cells might require updating - these operations result in the loss of dirty state, even though the outputs are preserved. If left as-is it will lead to mistakes.

![split-cell](https://user-images.githubusercontent.com/5832902/119876974-61775000-bf20-11eb-8071-fd9d48ef44bf.gif)

smacke commented:
Just tried it out and the stale indicator is awesome. I have another suggestion in the same vein as @krassowski's -- it seems that cut + paste does not preserve the dirty indicator. Not sure about other folks but I often cut + paste to move cells around; depending on how common of an operation it is, could be good to support. Last I checked cut + paste does not preserve the pre-cut cell's id; I didn't look at the stale indicator code in depth but I wouldn't be surprised if this is related.

martinRenou commented:
I'll see how I can implement keeping the dirty state

martinRenou commented:
It seems that the copy/paste, split and other cell actions cannot be implemented "properly" (i mean without ugly workarounds) without having the `isDirty` state part of nbformat, as it's the official cell JSON format that gets moved around.

How should we proceed? Would it be ok to wait for the `isDirty` state to be part of nbformat or are the workarounds fine?

krassowski commented:
> it seems that cut + paste does not preserve the dirty indicator. Not sure about other folks but I often cut + paste to move cells around; depending on how common of an operation it is, could be good to support

I agree. One could go a step further and say that some operations of moving cells change execution order and should add dirty state to all affected cells, this would be cell down from the change. This would not be the case if the cell was only moved past/in front of a markdown cell(s) though. This looks like something that would be useful and doable. Obviously this could be done in a subsequent PR.

goanpeca commented:
Agree on deferring to another PR. So far is a great start :)

In the meantime maybe create an issue to track the missing work and how it will depend on features added to nbformat?

martinRenou commented:
I moved the logic to the `CodeCellModel`, and made it a state change when the cell becomes dirty, I also added a test.

> In the meantime maybe create an issue to track the missing work and how it will depend on features added to nbformat

👍🏼

martinRenou commented:
Is there any more blocker for merging this?

krassowski commented:
I will do another review pass once binder is up again.

krassowski commented:
As mentioned earlier, I would prefer the indicator to be smart, this going away after reverting an accidental change and to add it in this PR; this way if it indeed causes performance issues we will be able to catch it in beta testing and adapt (by denouncing or doing something else). The handling of code cell split, copy and paste can be deferred to another PR.

As for performance I thought about hashing but it does not look like a great idea for JS as demonstrated in https://cs.stackexchange.com/questions/51933/string-comparison-vs-hashing - the overheads might be just too much compared to just comparing strings. However, we could have a simple conditional on the length of the contents (e.g. if the cell has more than 10^6 characters we defer or cancel the check).

martinRenou commented:
As you suggested @krassowski, I implemented the revert to clean state:

![revert](https://user-images.githubusercontent.com/21197331/120619861-ae957d80-c45c-11eb-8640-b25b8b659ad4.gif)

jtpio commented on packages/cells/src/model.ts:
```diff
@@ -807,19 +838,32 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
+    }
   }
 
+  isDirty = false;
```
How about making this a getter?

fcollonval reviewed (approved):
@martinRenou i updated this PR to handle the case of cleared output (+ some optimizations).

Before c754c73b94b5f691dc7972ddf28041a03674592c
![trouble_cleared_output](https://user-images.githubusercontent.com/8435071/121022597-c180b880-c7a2-11eb-9ec4-825f472c6d60.gif)

After
![correction_cleared_output](https://user-images.githubusercontent.com/8435071/121022826-00167300-c7a3-11eb-9406-effb37f222e0.gif)

It this is good for you, I'm :+1: with this one.

martinRenou commented:
Thanks!

jtpio commented on packages/cells/src/model.ts:
```diff
@@ -681,9 +686,35 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
+    // but for performance reason, the diff status is stored in a boolean.
+    return this._isDirty;
+  }
+  set isDirty(v: boolean) {
```
Do we need this setter to be public? Do we expect users of the model to be able to set the flag? Or could this be a private method?

martinRenou replied:
We need commands like split-cell to be able to set the dirty flag, so I am tempted to say yes, but I might not have the understanding of the whole thing.

jtpio replied:
Right, but will these commands manipulate the dirty state directly on the model?

martinRenou replied:
That's what I understood yes.

The annoying thing is that those commands manipulate the JSON model of the cell, the one that relates to what is actually loaded and saved in the Notebook file, at least that is what I understood, please correct me if I am wrong. So we might need the dirty state to be part of the official notebook format before we can implement this correctly.

jtpio replied:
Ah ok. I was searching the code from this branch for when the dirty state is set with `.isDirty = `, and it seemed to be coming from the model itself only. Which suggested having the setter be public might not be needed?

Because then we could have changed the attribute on the interface to be: `readonly isDirty: boolean;`

martinRenou replied:
Sure. It's fine making it readonly. We can change it later if needed :)

martinRenou commented:
@jtpio is it looking good in your opinion?

UI tests are not passing, but they are also failing on master (and looking at the diff it looks like a font rendering issue)

jtpio commented on packages/cells/src/model.ts:
```diff
@@ -103,6 +103,11 @@ export interface ICodeCellModel extends ICellModel {
+  /**
+   * Whether the code cell has been edited since the last run.
+   */
+  isDirty: boolean;
```
Should we make it `readonly`?

```suggestion
  readonly isDirty: boolean;
```

jtpio replied:
Which could then be moved above in the interface next to the other `readonly` attributes.

jtpio replied:
@martinRenou what do you think about this comment :point_up:

jtpio replied:
Done in https://github.com/jupyterlab/jupyterlab/pull/10296/commits/da20aeb0ec9e48f2453df2947369178e2f02298a

martinRenou replied:
Sorry I missed this comment 👀

jtpio commented on packages/cells/src/model.ts:
```diff
@@ -681,9 +686,36 @@ export class CodeCellModel extends CellModel implements ICodeCellModel {
+    return this._isDirty;
+  }
+
+  private setDirty(v: boolean) {
```
Let's rename with the `_` prefix like other private identifiers, and add a small docstring?

jtpio commented:
Thanks @martinRenou.

I left two more comments, but otherwise looks good :+1:

martinRenou commented:
Solved :)

jtpio commented:
> UI tests are not passing, but they are also failing on master (and looking at the diff it looks like a font rendering issue)

Opened https://github.com/jupyterlab/jupyterlab/issues/10371 to track this.

jtpio commented:
@martinRenou mind doing one more rebase to see if the UI tests pass, now that https://github.com/jupyterlab/jupyterlab/pull/10376 has been merged?

Also since this PR introduces a visual change, it could be interesting to add a new UI Test (with Galata) to cover it (with a new reference screenshot). Could be done in a follow-up PR though.

martinRenou commented:
I will add a visual regression test :)

jtpio commented:
Looks good!

![image](https://user-images.githubusercontent.com/591645/121716138-f8770700-cadf-11eb-8b7d-cc80fc366794.png)

jtpio reviewed (approved):
Thanks!
